web/src/message_view.js

```diff
@@ -661,6 +661,86 @@ export function show(raw_terms, opts) {
     }
 }
 
+function handle_anchor_case(anchor, has_found_anchor, get_msg_id) {
+    // These functions are scoped inside `handle_anchor_case` to
+    // to avoid them being used for any other purpose.
+    function duplicate_current_msg_list_with_new_data(data) {
+        const msg_list = new message_list.MessageList({data});
+        msg_list.reading_prevented = message_lists.current.reading_prevented;
+        return msg_list;
+    }
+
+    function select_msg_id(msg_id, select_opts) {
+        message_lists.current.select_id(msg_id, {
+            then_scroll: true,
+            from_scroll: false,
+            ...select_opts,
+        });
+    }
+
+    function select_anchor_using_data(data) {
+        const msg_list = duplicate_current_msg_list_with_new_data(data);
+        message_lists.update_current_message_list(msg_list);
+        // `force_rerender` is required to render the new data.
+        select_msg_id(get_msg_id(), {force_rerender: true});
+    }
+
+    if (has_found_anchor(message_lists.current.data.fetch_status)) {
+        select_msg_id(get_msg_id());
+    } else {
+        let msg_list_data = new MessageListData({
+            filter: message_lists.current.data.filter,
+            excludes_muted_topics: message_lists.current.data.excludes_muted_topics,
+        });
+        load_local_messages(msg_list_data);
+
+        if (has_found_anchor(all_messages_data.fetch_status)) {
+            select_anchor_using_data(msg_list_data);
+        } else {
+            msg_list_data = new MessageListData({
+                filter: message_lists.current.data.filter,
+                excludes_muted_topics: message_lists.current.data.excludes_muted_topics,
+            });
+            // We are fetching either `FETCH_SIZE` messages forward or backward from the anchor
+            // for currently handled anchors (oldest/newest). It is important that we
+            // are quick to make the jump feel fast while keeping enough context.
+            const FETCH_SIZE = 50;
+            message_fetch.load_messages({
+                anchor,
+                num_before: FETCH_SIZE,
+                num_after: FETCH_SIZE,
+                cont() {
+                    select_anchor_using_data(msg_list_data);
+                },
+                msg_list_data,
+            });
+        }
+    }
+}
+
+export function fast_track_current_msg_list_to_anchor(anchor) {
+    assert(message_lists.current !== undefined);
+    if (message_lists.current.visibly_empty()) {
+        return;
+    }
+
+    if (anchor === "oldest") {
+        handle_anchor_case(
+            anchor,
+            (fetch_status) => fetch_status.has_found_oldest(),
+            () => message_lists.current.first().id,
+        );
+    } else if (anchor === "newest") {
+        handle_anchor_case(
+            anchor,
+            (fetch_status) => fetch_status.has_found_newest(),
+            () => message_lists.current.last().id,
```

Comment:
I'm not sure I see why `get_msg_id` should be a callback, rather than just passing the number? 

I guess is the concern about what happens if new messages arrive after we hit `end` but before the server returns the data for that view? Oh, no, it's because `message_fetch.load_messages` will have added the new messages that it fetched. OK. So I think we just need a comment, maybe in the definition of the parameters in `handle_anchor_case`, explaining these parameters including why it's important that this one be a callback.

Might still be cleaner if what we passed was a `(msg_list_data) => msg_list_data.last().id` type callback, maybe named `message_list_data_to_target_message_id`, just to be more parallel to the API for the other one. Which could also be a more verbose name that spells out exactly what it does `fetch_status_shows_anchor_fetched(fetch_status)`

Reply:
Updated to do the renames.